src/shared/sleep-config.c

```diff
@@ -526,6 +529,90 @@ int get_total_suspend_interval(Hashmap *last_capacity, usec_t *ret) {
         return 0;
 }
 
+static int read_power_supply_battery_alarm_file(const char *alarm_filepath, int *ret) {
+        _cleanup_free_ char *batterytrippoint = NULL;
+        int battery_alarm, r;
+
+        r = read_one_line_file(alarm_filepath, &batterytrippoint);
+        if (r < 0)
+                return r;
+
+        r = safe_atoi(batterytrippoint, &battery_alarm);
+        if (r < 0)
+                return r;
+
+        *ret = battery_alarm;
+
+        return 0;
+}
+
+/* Return true if all batteries have acpi_btp support */
+int battery_trip_point_alarm_exists(void) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *dev;
+        int r, battery_count = 0, battery_alarm_count =0;
+
+        r = battery_enumerator_new(&e);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to initialize battery enumerator: %m");
+
+        FOREACH_DEVICE(e, dev) {
+                const char *battery_name, *alarm_filepath;
+                int battery_alarm;
+
+                r = sd_device_get_property_value(dev, "POWER_SUPPLY_NAME", &battery_name);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery name, ignoring: %m");
+                        continue;
+                }
+
+                alarm_filepath = path_join(POWER_SUPPLY_DIRPATH, battery_name, "alarm");
+                r = read_power_supply_battery_alarm_file(alarm_filepath, &battery_alarm);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery alarm from %s, ignoring: %m", alarm_filepath);
+                        continue;
+                }
```

Comment:
It seems not necessary to read battery name and build path to sysfs attribute. I guess the following should be simpler:
```
const char *s;
int a;

r = sd_device_get_sysattr_value(dev, "alarm", &s);
if (r < 0) {
        log_device_debug_errno(...);
        continue;
}

r = safe_atoi(s, &a);
if (r < 0)
        ...
```

Reply:
Yeah makes sense. Will fix.